services/api/service.go

```diff
@@ -1778,6 +1785,13 @@ func (api *RelayAPI) handleSubmitNewBlock(w http.ResponseWriter, req *http.Reque
 		}
 	}
 
+	// Refuse blocks outside the specified gas limit
+	if payload.Capella.ExecutionPayload.GasLimit > RelayActualGasLimit {
+		return
+	}
+	// Overwrite the builder's gasLimit with the relay-set fictitious limit
+	payload.Capella.ExecutionPayload.GasLimit = RelayFictitiousGasLimit
```

Comment:
overriding any payload field renders the payload's block hash invalid and the payload simulation would fail on this integrity check. might be missing smth obv here...
https://github.com/flashbots/builder/blob/ca9f4b1ae44aadc6c254e9198ce72d143c19badd/eth/block-validation/api.go#L247-L249

Reply:
ya there are 2 changes to be made rn

- change `utils.CalcGasLimit` to not take the validator preference
- Set payload gasLimit to output of `utils.CalcGasLimit`
- Calculate correct hash & override payload header hash

Reply:
okay new implementation plan:

We require the builder backend to look up the previous blockhash in order to calculate the correct gasLimit given the ficitious registered target. We then require a way to get that info back to the relay API. So we need to return data from the builder's RPC endpoint that currently returns no data. That data is the new gas_limit and blockhash. This output can then used to modify the Builder's payload 

So to enumerate changes:
- modify the [ValidateBuilderSubmissionV2](https://github.com/flashbots/builder/blob/ca9f4b1ae44aadc6c254e9198ce72d143c19badd/eth/block-validation/api.go#L231) behavior to calculate the correct gasLimit and blockhash before simulating the block
- modify ValidateBuilderSubmissionV2 to return the modified gasLimit and blockHash
- modify the relay API to override the builder payload with the returned result before storing it in the bids

Reply:
okay updated 👍